experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,906 @@
+    index_length: u32,
+    data_length: u32,
+    high_start: u32,
+    shifted12_high_start: u16,
+    index3_null_offset: u16,
+    data_null_offset: u32,
+    null_value: u32,
+}
+
+// CPTAuto will only expose get_u32() (probably will call it `get()`), but will not / cannot expose get() without using an enum for W
+// CPT<W1,S1> will expose get() and get_u32(). You may not want get_u32() at this point, but it is necessary in order to implement CPTAuto
+
+// u32 vs. char in Rust - unpaired surrogates are valid code points but are not UTF-32 code units, therefore not allowed in `char` type in Rust.
+// But for the CodePointTrie, we do want to be able to look up surrogate code points, therefore we want to use the u32 type. (Note: this is also what we do in the `uniset` crate for `UnicodeSet`.)
+// struct CodePoint(u32) ← enforce whatever invariants you'd like
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    fn try_new(
+        header: CodePointTrieHeader,
+        index: &[u16],
+        data: &[W],
+    ) -> Result<CodePointTrie<'trie, W, T>, Error> {
+        if header.data_length < ERROR_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized(FromDeserializedError {
+                reason: "Data array must be large enough to contain error value",
+            }));
+        }
+
+        if header.data_length < HIGH_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized(FromDeserializedError {
+                reason:
+                    "Data array must be large enough to contain value for range highStart..U+10FFFF",
+            }));
+        }
+
+        if index.len() as u32 != header.index_length {
+            return Err(Error::FromDeserialized(FromDeserializedError {
+                reason: "Length of index array does not match corresponding header value",
+            }));
+        }
+
+        if data.len() as u32 != header.data_length {
+            return Err(Error::FromDeserialized(FromDeserializedError {
+                reason: "Length of data array does not match corresponding header value",
+            }));
+        }
+
+        // TODO: Why does this check not work on the test data?
+        // Is it invalid generally, and if so, why?
+        //
+        // if (index.len() as u32) < T::FAST_MAX {
+        //     return Err(Error::FromDeserialized(FromDeserializedError{
+        //         reason: "Length of index array does not have enough entries to support 2-step lookups (range 0..fastMax) for this trie type"}));
+        // }
+
+        // TODO: what are other assertions to include?
+
+        // Note: this particular constructor is "templatized" through Rust's
+        // generics type system, so callers to this constructor function must
+        // necessarily know ahead of time which trie type and value width their
+        // desired trie instance represents. There is no runtime matching ->
+        // dynamic dispatch here.
```

Comment:
Suggestion: This seems like it should go in a doc comment on the function, rather than an inline comment in the code.